company-tng.el

```diff
@@ -133,25 +126,17 @@ confirm the selection and finish the completion."
 (defvar company-rtags-insert-arguments)
 (defvar lsp-enable-snippet)
 
+(defgroup company-tng nil
+  "Company Tab and Go."
+  :group 'company)
+
+(defcustom company-tng-auto-configure t
+  "Automatically apply default configure when enable `company-tng-mode'."
+  :type 'boolean)
```

Comment:
When `company-tng-auto-configure` is `t`, enabling and then immediately disabling `company-tng-mode` leaves undoable side-effects on the configuration, which affects company in general. Usually I expect the effects of a minor mode to go away after I disable it.
I wonder if it would be better to leave this `nil`, and then set it to `t` if the user calls `company-tng-configure-default`, since that function implies that the user wants the extra configuration.
I don't have a strong opinion (since I also see benefits for leaving it `t` by default), I'm just wondering what you guys think about it.

Reply:
There's also an option of saving the previous configuration somewhere, and restoring it when the mode is disabled.

Reply:
Regarding the bindings specifically, maybe it could replace `company-active-map` wholesale with a tng-specific map (which inherits from it, of course). And restore it, of course, when the mode is disabled.

Reply:
Not sure if it's really worth going this far... It's not like `company-tng` is toggled on and off frequently or as part of normal usage. Also, won't a new map break existing users of `company-tng` in case they customized key-bindings?

Reply:
Since we specifically mentioned `company-tng-configure-default` is an obsoleted alias of `company-tng-mode`, I think this should be `t` by default.
If most of `company-tng` user is already use `company-tng-configure-default`, I think this only affect the minimal number of user who manually add `company-tng-frontend` only. Even in that case, setting `company-tng-auto-configure` to `nil` will solve this problem.
Perhaps we can do better document so user aware of the auto configure stuff?

Reply:
But what's the use case of setting it to `nil`? Redoing the whole configuration manually?

If the user is willing to do that, they could as well change `company-frontends` as appropriate and set `company-selection-default` to nil, too. If `company-tng--allow-unselected` becomes unnecessary, not much else will be needed. That would only leave Eglot requiring an advice, but I'd be happy to direct anybody complaining about that to Eglot's issue tracker.

Reply:
> But what's the use case of setting it to `nil`? Redoing the whole configuration manually?

Not necessarily everything, but only needed thing.
For example, I'm more favorable of using snippets with `lsp-mode` than just a few glitch on the overlay preview is not displayed correctly, so I'm not setting any `lsp-enable-snippet` to `nil`.
The end result is that I still got the `tab and go` functionality of selecting a candidate and then can type the next one immediately, with the cost of some visual glitch that can easily ignored.


Also, I think that `company-selection-default` should only be for `company` plugin package to use only, user should not set that. Since setting `company-selection-default` without a frontend that know how deal with it.
For example:
- `company-tng` is showing some cues in form of overlay.
- Or a new front end that can support multiple actions on selecting a candidate, like: pressing <Enter> when there's no selection will just insert new line, but if there's selection, it will show a popup menu asking what user want to do with this candidate. In that case, the new frontend can set `company-selection-default` to `nil` and benefit from that.

The `company-tng--alow-unselected` is tricky, since only `tng` show overlay preview and allow user to select and commit the candidate immediately, thus it will need the `allow-unselected` to go to unselected state from selected. The imagine frontend that I took as example a few lines ago will not need this functionality.

Reply:
> I'm more favorable of using snippets with lsp-mode than just a few glitch on the overlay preview is not displayed correctly, so I'm not setting any lsp-enable-snippet to nil

But if you disable the auto-configuration, you will have to include the rest of the configuration in your config, won't you? That's the "everything" that I meant.

Whereas if `company-tng-mode` always did its whole thing, you could just add a short function to `company-tng-mode-hook` (for example) to tweak the 1-2 values you want to be different.

> I think that company-selection-default should only be for company plugin package to use only

That also weighs in favor of the approach described above.

> but if there's selection, it will show a popup menu

Frontends normally should limit themselves to visualization, at least while there is no good way to implement something like this in a frontend. But OK, let's consider this potentiality as well.

> The imagine frontend that I took as example a few lines ago will not need this functionality.

But will it be hindered by it?

Reply:
> But if you disable the auto-configuration, you will have to include the rest of the configuration in your config, won't you? That's the "everything" that I meant.

The large part of `company-tng-configure-default` are settings external package variables and add new key bindings to `company-active-map`. If I'm happy with the old key bindings and don't want external package variables to be set, isn't it easier to just disable the auto configuration and no need to do anything else?
The `company-tng-frontend` is always added into `company-fronends` list, regardless of auto configuration or not.

Before user can easily ignore all auto configuration stuff by do not call `company-tng-configure-default`. Now I think we should also provide alternative method to do that without much trouble of reset everything in `company-tng-mode-hook`.



> > The imagine frontend that I took as example a few lines ago will not need this functionality.
> 
> But will it be hindered by it?

Probably not, since the new imaginary frontend is not having functionality of auto commit the selected candidate.

To be honest, merging the `company-tng--alow-unselected` with `company-select-next` is tricky and not that clean, since it has to check for `company-selection-default` value and act based on that. Other command no need to do that and only need to check for `company-selection` value to be `nil` or not.

Reply:
> If I'm happy with the old key bindings and don't want external package variables to be set, isn't it easier to just disable the auto configuration and no need to do anything else?

Seems like I misunderstood you, and you not only want to keep the snippet expansion as-is, but keep the default key bindings as well? If we do expect this to be a significant use case for `company-tng-mode` users, then yes, you guys' suggestion to add the `-configure-default` variable and default it to `t` sounds good.

The function that would be called when this variable is `t` should be called something else, though. And `company-tng-configure-default` should be an obsolete alias for `company-tng-mode` to ease the transition.

> To be honest, merging the company-tng--alow-unselected with company-select-next is tricky and not that clean, since it has to check for company-selection-default value and act based on that.

Right, but if `company-selection` being nil is a valid core behavior now, why wouldn't it check that? And we would benefit from having a test for this.

Reply:
> Right, but if `company-selection` being nil is a valid core behavior now, why wouldn't it check that

I mean checking to make sure `company-select-next` is behave correctly when `company-selection` is `nil` is already integrated, and quite straight forward.
However, the part of make `company-selection` to become `nil` again from other value in `company-select-next` is not clean and too dependent on `company-tng`.



> And we would benefit from having a test for this.

I'm working on adding test for this

Reply:
@kiennq How is `company-selection` transitioning to `nil` is more dependent on `company-tng` than transitioning from `nil`?
If the latter is implemented inside core company, I agree with @dgutov that it makes sense to implement the former, for symmetry. And then get rid of the advice altogether.

Reply:
@nikital It's because while `company-selection` from `nil` is quite straight forward, the transition to `nil` required checking value of `company-selection-default`, `company-selection-wrap-around` to determine the outcome.
For the details:
- Currently `company-selection-next` just calculate the selection and pass to `company-set-selection` which will calculating `company-selection-wrap-around` to correct index and set it. (1)
- If we want to support transition to `nil`, we will have to check the value of `company-selection-default` (which is not checked in any function, we only check value of `company-selection`), calculating the desired selection regarding to `company-selection-wrap-around` and pass it to `company-set-selection`. (2)

Trying to force (2) in `company-selection-next` make the function become more bulky and not clean at all.

Since `company-selection` can be `nil` we have to handle the case of transition from `nil` anyway, but the case of transition to `nil` is kind of `company-tng` specialty, since without it, `company-tng` cannot easily cancel the completion once selected.
The other front end may have `company-selection-default` to `nil` but if it doesn't have same behavior of committing selection immediately like `company-tng`, it will not require transition to `nil`.

Reply:
@kiennq I think I see your point, if you're saying that `company-tng` currently doesn't obey `company-selection-wrap-around`, and it seems messy to fix that. I'm not going to insist you do that as part of this PR.

But the mess is here already (again: `company-tng` doesn't obey a documented user option), and the conflicts will need to be resolved anyway. Keeping the behavior in an advice doesn't make the whole body of code less "bulky" and "not clean".

So, up to you, but it would be great to choose a behavior that honors both tng and wrap-around. Yes, it will have to check the value of `company-selection-default`. But it's a variable from the same package, so there's no ugliness in that.

Reply:
Okay, I've added test and also remove the `advice-add`. It should be better now

Reply:
Thanks!

What's your situation with copyright assignment to FSF?

Reply:
I've completed it.